### ISSUE TITLE
(core): Property value of a resource inside a CDK Stage contains the CDK Pipelines pipeline construct ID

### ISSUE THREAD

**Issue by alexpulver** [aws/aws-cdk#16830]
### What is the problem?

Per @rix0rrr, logical IDs generated by the AWS CDK should be based on the construct path to the nearest containing stack. Currently, this logic doesn't hold for resource property names auto-generated by the AWS CDK for resources that MUST have a name in CloudFormation.

Consider the following construct tree and resources:  `App` -> `Pipeline stack (CDK Pipelines)` -> `Prod stage` -> `Stateless stack`-> `Amazon CodeGuru Profiler profiling group`. The current implementation adds the pipeline stack construct ID as a prefix for the profiling group name. That creates an unneeded dependency between pipeline construct ID and production resources. Changing pipeline construct ID can affect production resources (e.g. replacement).

### Reproduction Steps

```bash
git clone  https://github.com/alexpulver/usermanagement-backend
cd usermanagement-backend/
export AWS_ACCESS_KEY_ID="***"
export AWS_SECRET_ACCESS_KEY="***"
export AWS_SESSION_TOKEN="***"
python3.7 -m venv .venv
source .venv/bin/activate
./scripts/install-deps.sh
npx cdk ls "**"
grep ProfilingGroupName cdk.out/assembly-UserManagementBackend-Pipeline-UserManagementBackend-Prod/UserManagementBackendPipelineUserManagementBackendProdStateless5E346530.template.json
```

The Amazon CodeGuru Profiler profiling group name below contains `UserManagementBackendPipeline` string, that is generated from the construct ID of the pipeline stack:
```
"ProfilingGroupName": "UserManagementBackendPipelineUserManagementBackendProdStatelessAPILambdaFunctionProfilingGroup9415F6C1",
```

### What did you expect to happen?

I would expect that profiling group name will start from the construct ID of the nearest containing stack:
```
"ProfilingGroupName": "UserManagementBackendProdStatelessAPILambdaFunctionProfilingGroup9415F6C1",
```

### What actually happened?

The Amazon CodeGuru Profiler profiling group name below contains `UserManagementBackendPipeline` string, that is generated from the construct ID of the pipeline stack:
```
"ProfilingGroupName": "UserManagementBackendPipelineUserManagementBackendProdStatelessAPILambdaFunctionProfilingGroup9415F6C1",
```

### CDK CLI Version

1.117.0 (build 0047c98)

### Framework Version

1.117.0

### Node.js Version

v14.16.1

### OS

macOS Catalina 10.15.7

### Language

Python

### Language Version

Python (3.7.9)

**Comment by moltar** [comment]
Thanks, @alexpulver for filing this issue!

FYI, this comes from a thread from the [CDK Slack chat](https://cdk-dev.slack.com/archives/C01FDB9UFNV/p1633502785198500).

I feel like this is a classic [action at a distance](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming)) antipattern and it should not be happening.

The following resources (but not limited to), seem to be affected by the pipeline name injection:

- Custom::DatabaseMigration
- AWS::ECS::TaskDefinition
- AWS::Lambda::Permission
- AWS::Lambda::EventSourceMapping
- AWS::Events::EventBus
- AWS::ApiGatewayV2::ApiMapping

**Comment by kennu** [comment]
I am experiencing this problem when I want to deploy the same stack both manually (`cdk deploy`) and using [CodePipeline](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_pipelines.CodePipeline.html). Because the generated identifiers change when the Stack is inside a Stage, which is inside a Pipeline Stack, there is always a new deployment. I would like to find a way to use CodePipeline without adding the pipelinestack/stage/ prefix to all the identifiers, so that the stack is identical regardless of how it is deployed.

**Comment by ssaarela** [comment]
We ran into this issue with multi-branch pipelines with some shared resources/stacks like networking and database. A shared stack is developed in a feature branch with its own pipeline and after a merge the same stack is deployed by the main pipeline all the way through preprod to prod. This does't really work as the Logical IDs change after the merge and some resources are needlessly attempted to be replaced.

The way a stack is deployed, either manually or from one pipeline or another (e.g. main vs hotfix pipeline), should have no effect on the generated resources whatsoever. This bug creates a bi-directional dependency from actual resources to the pipeline which deployed them, making the pipeline effectively the owner of the deployed stacks and in worst case prevents updates form other sources.

In addition to the resources @moltar mentioned, this issue also affects SecurityGroups and generated DatabaseSecrets, which are not that easily replaced.

**Comment by MikeMather** [comment]
Any update on this? I've moved our deployments into a pipeline and am now running into the same issue. The deployment pipeline name is prefixing all of the resource names causing it to try to replace tons of existing resources which is causing issues.

**Comment by OperationalFallacy** [comment]
Still prefixing it

We managed to deploy a bunch of stacks using custom `stackName` in the constructor

```
  constructor(scope: Construct, id: string, props: DBPersistenceStackProps) {
    super(scope, id, {
      stackName: `db-${props.environmentName}`,
      ...props,
    });
```

However, it is trying to change ids on specific resources like security groups (the description, when absent), SecurityGroupIngress, subnets and some others 

```
[~] AWS::EC2::SecurityGroup dev/db-dev/DBSecurityGroup DBSecurityGroupE3B245A3 replace
 ├─ [~] GroupDescription (requires replacement)
 │   ├─ [-] db-dev/DBSecurityGroup
 │   └─ [+] pipeline-cicd/dev/db-dev/DBSecurityGroup
```

Is there a way to override it? Its

**Comment by l3ku** [comment]
Also faced this issue when moving some stack definitions from `app.py` to a `Stage`, where the stacks had already been deployed when they were defined in `app.py`. The stage name is now included in the logical ID of some resources, and hence CFN tries to re-create the resources due to the logical ID change. There doesn't seem to currently be any other workaround I can think of apart from hard-coding the logical IDs or redeploying the stack.

**Comment by moltar** [comment]
This is really biting us! We have performed an initial deployment of a very complex application (account-based multitenancy) with many stacks, from local. And then connected it to an existing pipeline that already deploys several other tenants. And, everything is failing, because resources, and thereby exports, are being fully replaced.

The only solution is to nuke everything and start over. This is a day's worth of work.

**Comment by moltar** [comment]
I think I found a relatively clean workaround.

If you create an instance of `Stage` with `scope` being the instance of `App` (root), and use that inside the pipeline, then the pipeline works just fine, and the prefixes are no longer added, as the stage is scoped _not_ to the pipeline.

In my case, the pipeline was added to the root `App` scope as well, so I could just use the first positional scope value.

So, basically do this, even when inside the pipeline stack:

```ts
new Stage(app, ...) // where `app` is the instance of `App` (root)
```

Instead of:

```ts
new Stage(this, ...) // where `this` is an instance of `Stack` where your `CodePipeline` lives.
```

**Comment by ssaarela** [comment]
> I think I found a relatively clean workaround.
> 
> If you create an instance of `Stage` with `scope` being the instance of `App` (root), and use that inside the pipeline, then the pipeline works just fine, and the prefixes are no longer added, as the stage is scoped _not_ to the pipeline.
> 
> In my case, the pipeline was added to the root `App` scope as well, so I could just use the first positional scope value.
> 
> So, basically do this, even when inside the pipeline stack:
> 
> ```ts
> new Stage(app, ...) // where `app` is the instance of `App` (root)
> ```
> 
> Instead of:
> 
> ```ts
> new Stage(this, ...) // where `this` is an instance of `Stack` where your `CodePipeline` lives.
> ```

As @moltar says, this is a pretty clean workaround for pipeline name affecting logical IDs in stacks. Unfortunately that still leaves a Stage name to be included in logical IDs and the same trick doesn’t work with Stages. It's not quite as bad as having the pipeline name there, but it still ties stacks tightly with how (or from where) they’re deployed.

Is there a way of adding stacks as actions to a Stage without having them as it’s children, the same way a pipeline can have stages that are not it’s direct children?

Can logical ids be fixed using an aspect or something?

Is there any change getting this very issue fixed any time soon? Logical IDs are an internal concept of a CloudFormation Stack and nothing coming from outside should affect them. 

This is very important for logical ID stability and as the doc says:  > Avoid changing the logical ID of a resource after it has been created. AWS CloudFormation identifies resources by their logical ID. Therefore, if you change the logical ID of a resource, AWS CloudFormation creates a new resource with the new logical ID, then deletes the existing one. Depending on the type of resource, this might cause service interruption, data loss, or both.

**Comment by moltar** [comment]
This coming back to bite us, again and again.

I am on a new project where stages were not previously used. Adding a Stage into the mix, even with hard-coded stack names (via `stackName` prop), still leaves _tons_ of renames, including resource names, exports, logical IDs, etc... It's not going to deploy cleanly at this point, without resource replacements.

This is blocking us from adopting the concept of stages thereby making us unable to use the CDK Pipelines, and thereby blocking us from using the CodePipeline service.
